### PR TITLE
feat: use composition for kernel approximation

### LINKED ISSUE #297: Rename approximation._anchor_body or move into ANNchorApproximator
Should be called `_annchor_body`. Better housekeeping would be to include it as a static method in `ANNchorApproximator`, although care needs to be taken that JIT still works.

### LINKED ISSUE #544: Use composition for kernel `aproximators` like in `SteinKernel`
### What's the new feature?

Allow kernel approximation on all kernel methods by utilising composition (like in `SteinKernel`); a `CompositeKernel` base could be shared across such kernels (including `SteinKernel`).
```python
# Current implementation
class Kernel:
    def approximate_kernel_matrix_row_sum_mean(self, ...):
        self.approximator.approximate(self)

class KernelMeanApproximator:
    def approximator(self, kernel):
        ...  # Compute the approximate kernel matrix row sum mean

# Proposed implementation
class ApproximateKernel(coreax.kernel.Kernel):
    def __init___(self, ..., base_kernel):
        self.base_kernel = base_kernel
        ...

    @override
    def calculate_kernel_matrix_row_sum_mean(self, ...):
        ...  # Compute the approximate kernel matrix row sum mean using `self.base_kernel`
```

### What value does this add?

Allows downstream users of `Kernel` to be freed of the responsibility of handling approximation (responsibility is solely that of the composed kernel). This additionally allows for approximation of other aspects of a `base_kernel` and makes clear what method is being approximated (making the code more self descriptive).

This also complements the changes proposed in #534, allowing `refine` to be blind to the use of approximation in the calculation of the `kernel_matrix_row_sum`.

### Is there an alternative you've considered?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tc85324**
Closes #544, Closes #297

### PR Type

- Feature
- Code style update (formatting, local variables)
- Refactoring (no functional changes)
- Documentation content changes
- Tests

### Description
- Update `kernel.py` and `approximation.py` module docstrings
- Use `@override` to reduce docstring reuse
- Use `equinox.Module` for all `Kernel`s
- Remove now redundant approximator handling from `coresubset.py`, `kernel.py`, and `refine.py`
- Remove now redundant/incompatible tests from `test_refine.py`, `test_kernel.py`, `test_coresubset.py` and `test_approximation.py`
- Update `herding_approximate_gram_matrix.py` example
- Disable redundant/incompatible pylint `abstract-methods` rule
- Add ruff-pylint exception for `__check_init__` dunder method
- Fixes typo `_anchor_body` -> `_annchor_body`

### How Has This Been Tested?
All the tests pass.

### Does this PR introduce a breaking change?
Yes, the kernel approximation API has completely changed (there is no feasible deprecation strategy).

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Review by pc532627 — CHANGES_REQUESTED** [review]
Only very minor points. Looking at historic test workflows, as far as I can tell, there is no performance impact on the code

File: coreax/approximation.py [hunk 1]
```diff
@@ -13,419 +13,263 @@
...
-above calculations, without having to perform as many evaluations of the kernel. All
-approximators in this module implement the base class :class:`KernelMeanApproximator`.
+Classes and associated functionality to approximate kernels.
+
+When a dataset is very large, methods which have to evaluate all pairwise combinations
+of the data, such as :meth:`~coreax.kernel.Kernel.calculate_kernel_matrix_row_sum_mean`,
+can become prohibitively expensive. To reduce this computational cost, such methods can
+instead be approximated (providing suitable approximation error can be achieved).
+
+The :class:`ApproximateKernel`\ s in this module provide the functionality required to
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/approximation.py:23]: This provides a good explanation https://stackoverflow.com/a/23853904, there are other options but I had opted for this as it was the most concicse

File: coreax/approximation.py [hunk 2]
```diff
@@ -13,419 +13,263 @@
...
+import jax.tree_util as jtu
+from jax import Array
 from jax.typing import ArrayLike
+from typing_extensions import Literal, override
+
+from coreax.kernel import CompositeKernel, Kernel
+from coreax.util import KeyArrayLike
+
+
+def _random_indices(
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/approximation.py:44]: I wasn't going to document these as they are marked private (by the underscore), and their functionality is reasonably self descriptive, but will included if you think its important to have.

File: coreax/approximation.py [hunk 3]
```diff
@@ -13,419 +13,263 @@
...
+    def divergence_x_grad_y_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
+        return self.base_kernel.divergence_x_grad_y_elementwise(x, y)
 
-        The parameter `num_kernel_points` can take any non-negative value, however,
-        setting this to 0 will simply produce a kernel mean approximation of zero at all
-        points.
+
+class RandomRegressionKernel(Kernel):
+    """
+    An kernel that requires the attributes for random regression.
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/approximation.py:114]: Good spot!

File: coreax/approximation.py [hunk 4]
```diff
@@ -13,419 +13,263 @@
...
-    .. note::
+    @override
+    def divergence_x_grad_y_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
+        return self.base_kernel.divergence_x_grad_y_elementwise(x, y)
 
-        The parameter `num_kernel_points` can take any non-negative value, however,
-        setting this to 0 will simply produce a kernel mean approximation of zero at all
-        points.
+
+class RandomRegressionKernel(Kernel):
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/approximation.py:112]: Correct, this is mean't to be abstract still, it allows an "abstract" type hierarchy

File: coreax/approximation.py [hunk 5]
```diff
@@ -13,419 +13,263 @@
...
-        # Randomly select training points
-        try:
-            train_idx = random.choice(
-                self.random_key,
-                num_data_points,
-                (self.num_train_points,),
-                replace=False,
+        features = features.at[:, 0].set(self.base_kernel.compute(data, data[0])[:, 0])
+
+        def _annchor_body(idx: int, features: Array) -> Array:
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/approximation.py:210]: The JIT would have been redundant as it is only called by `fori_loop`, which is already compiling anyway

File: coreax/coresubset.py [hunk 6]
```diff
@@ -136,40 +129,29 @@ def tree_flatten(self) -> tuple[tuple, dict]:
...
     def fit_to_size(self, coreset_size: int) -> None:
         r"""
         Execute kernel herding algorithm with Jax.
 
-        We first compute the kernel matrix row sum mean (either exactly, or
-        approximately) if it is not given, and then iterative add points to the coreset
-        balancing  selecting points in high density regions with selecting points far
-        from those already in the coreset.
+        We first compute the kernel matrix row sum mean if it is not given, and then
+        iterative add points to the coreset balancing  selecting points in high density
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/coresubset.py:140]: Well spotted!

File: coreax/coresubset.py [hunk 7]
```diff
@@ -212,7 +194,6 @@ def fit_to_size(self, coreset_size: int) -> None:
         self.coreset = self.original_data.pre_coreset_array[self.coreset_indices, :]
 
     @staticmethod
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/coresubset.py:196]: The JIT is redundant as the function is only called by the `fori_loop`, which is already compiling anyway

File: coreax/coresubset.py [hunk 8]
```diff
@@ -552,7 +533,6 @@ def fit_to_size(self, coreset_size: int) -> None:
...
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/coresubset.py:535]: Same as the other responses

File: coreax/kernel.py [hunk 9]
```diff
@@ -24,56 +24,45 @@
...
-:class:`~coreax.score_matching.ScoreMatching` to side-step this requirement.
-
-The other mandatory method to implement when defining a new kernel is
-:meth:`~coreax.kernel.Kernel.tree_flatten`. To improve performance, kernel computation
-is JIT compiled. As a result, definitions of dynamic and static values inside
-:meth:`~coreax.kernel.Kernel.tree_flatten` ensure the kernel object can be mutated and
-the corresponding JIT compilation does not yield unexpected results.
+    k(x,y) = \text{output_scale} * \exp (-||x-y||^2/2 * \text{length_scale}^2),
+
+which if parameterized with an ``output_scale`` of
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/kernel.py:34]: I'm not sure what defines British English, but [OED has it with a 'z'](https://www.oed.com/dictionary/parameterize_v)

File: coreax/approximation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pc532627) [reply coreax/approximation.py:23]: Thanks for the explanation - happy to keep it

File: coreax/approximation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pc532627) [reply coreax/approximation.py:44]: I think it is important to, please push updates for both

File: coreax/approximation.py [hunk 10]
```diff
@@ -13,419 +13,263 @@
...
+        if num_select > num_data_points:
+            raise ValueError(
+                f"'num_{mode}_points' must be no larger than the number of points in "
+                "the provided data"
+            ) from exception
+        raise
+    return selected_indices
+
+
+def _random_regression(
```
[the remark this answers is not in the document]
Comment (pc532627) [reply coreax/approximation.py:67]: I think it is important to, please push updates for both

File: coreax/kernel.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (pc532627) [reply coreax/kernel.py:34]: In this case lets leave it, sorry for the misflag!

**Comment by tc85324** [comment]
All issues should now be addressed. Additional changes have been made to allow for a successful merge. In particular, a test has been removed from Stein Thinning which cannot be made compatible with the new kernel design.

File: coreax/approximation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pc532627) [reply coreax/approximation.py]: The entry `features` in the docstring needs updating to `_features`

**Review by pc532627 — CHANGES_REQUESTED** [review]
Only one tiny change - otherwise all looks good @tc85324

File: coreax/approximation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pc532627) [reply coreax/approximation.py]: @tc85324  just for reference this is the single change needed

**Pull request merged**